Possible bug with timeseries gradients with RecursiveCheckpointAdjoint?

```
import jax.numpy as jnp
import jax

jax.config.update('jax_enable_x64', True)
import diffrax
import equinox
from statsmodels.tools import numdiff

import numpy.testing as np_testing
def simple_diffrax_dynamics_for_test(t, y, p):
  """Extremely simple dynamics function for tests."""
  return jnp.array([2 - p[0] * y[0], y[0] - p[1] * y[1], 0.])

initial_parameter_vector = jnp.array([2.5, 4.5, 2.5])

# Manually define a diffrax integrator:

def diffrax_loss_fcn(p):
  t = jnp.array([0., 0.5, 0.75, 20.])
  term = diffrax.ODETerm(simple_diffrax_dynamics_for_test)

  solver = diffrax.Kvaerno5(
      nonlinear_solver=diffrax.NewtonNonlinearSolver(
          atol=1e-8, rtol=1e-8, max_steps=15),)

  saveat = diffrax.SaveAt(ts=t)
  stepsize_controller = diffrax.PIDController(rtol=1e-8, atol=1e-8)

  sol = diffrax.diffeqsolve(
      term,
      solver,
      t0=t[0],
      t1=t[-1],
      y0=jnp.array([1., 1., 2.5]),
      saveat=saveat,
      dt0=None,
      stepsize_controller=stepsize_controller,
      max_steps=5000,
      args=p,
      throw=False,
  )
  return jnp.sum(jnp.square(sol.ys - jnp.ones_like(sol.ys)))

grad_of_loss_fn = jax.grad(diffrax_loss_fcn)
jitted_grad_of_loss_fn = jax.jit(grad_of_loss_fn)

diffrax_grad = grad_of_loss_fn(initial_parameter_vector)
jitted_diffrax_grad = jitted_grad_of_loss_fn(initial_parameter_vector)

numdiff_epsilon = (
    jnp.sqrt(jnp.finfo(diffrax_grad.dtype).eps) *
    jnp.maximum(jnp.abs(initial_parameter_vector), 0.1))
expected = numdiff.approx_fprime(
    initial_parameter_vector,
    diffrax_loss_fcn,
    epsilon=numdiff_epsilon)

np_testing.assert_allclose(diffrax_grad, expected, rtol=1.e-3, atol=1.e-3)
```

Gives this:

```
Mismatched elements: 2 / 2 (100%)
Max absolute difference: 3.82944562
Max relative difference: 10.26014782
 x: array([4.390821, 3.167287])
 y: array([0.561375, 0.281283])
```

and:
```
np_testing.assert_allclose(jitted_diffrax_grad, expected, rtol=1.e-3, atol=1.e-3)
```

gives this:
```
AssertionError: 
Not equal to tolerance rtol=0.001, atol=0.001

Mismatched elements: 2 / 3 (66.7%)
Max absolute difference: 3.82944562
Max relative difference: 10.26014782
 x: array([4.390821, 3.167287, 0.      ])
 y: array([0.561375, 0.281283, 0.      ])
```

--- a/diffrax/__init__.py
+++ b/diffrax/__init__.py
@@ -81,4 +81,4 @@ from .term import (
 )
 
 
-__version__ = "0.0.2"
+__version__ = "0.0.3"

--- a/diffrax/integrate.py
+++ b/diffrax/integrate.py
@@ -622,8 +622,8 @@ def diffeqsolve(
     if is_sde(terms):
         if not isinstance(solver, (AbstractItoSolver, AbstractStratonovichSolver)):
             warnings.warn(
-                f"`{solver.__name__}` is not marked as converging to either the Itô "
-                "or the Stratonovich solution."
+                f"`{type(solver).__name__}` is not marked as converging to either the "
+                "Itô or the Stratonovich solution."
             )
         if isinstance(adjoint, BacksolveAdjoint):
             if isinstance(solver, AbstractItoSolver):

--- a/diffrax/misc/__init__.py
+++ b/diffrax/misc/__init__.py
@@ -1,9 +1,4 @@
-from .ad import (
-    fixed_custom_jvp,
-    is_perturbed,
-    nondifferentiable_input,
-    nondifferentiable_output,
-)
+from .ad import fixed_custom_jvp, nondifferentiable_input, nondifferentiable_output
 from .bounded_while_loop import bounded_while_loop, HadInplaceUpdate
 from .errors import branched_error_if, error_if
 from .misc import (

--- a/diffrax/misc/ad.py
+++ b/diffrax/misc/ad.py
@@ -10,17 +10,17 @@ import jax.lax as lax
 from ..custom_types import PyTree
 
 
+# TODO: this will sometimes return False on a perturbed array, see JAX issue #9567.
+# Correspondingly it should *not be used* until that is fixed.
+# (The only use is in nondifferentiable_input, below, which will simply not raise
+# errors quite as frequently as it should do -- not too bad.)
 def is_perturbed(x: Any) -> bool:
-    return eqx.is_inexact_array(x)
-    # TODO:
-    # Use some more fine criterion, once JAX issue #9567 is fixed.
-    #
-    # if isinstance(x, jax.ad.JVPTracer):
-    #     return True
-    # elif isinstance(x, jax.core.Tracer):
-    #     return any(is_perturbed(attr) for name, attr in x._contents())
-    # else:
-    #     return False
+    if isinstance(x, jax.ad.JVPTracer):
+        return True
+    elif isinstance(x, jax.core.Tracer):
+        return any(is_perturbed(attr) for name, attr in x._contents())
+    else:
+        return False
 
 
 def nondifferentiable_input(x: PyTree, name: str) -> None:

--- a/diffrax/misc/ad.py
+++ b/diffrax/misc/ad.py
@@ -11,12 +11,16 @@ from ..custom_types import PyTree
 
 
 def is_perturbed(x: Any) -> bool:
-    if isinstance(x, jax.ad.JVPTracer):
-        return True
-    elif isinstance(x, jax.core.Tracer):
-        return any(is_perturbed(attr) for name, attr in x._contents())
-    else:
-        return False
+    return eqx.is_inexact_array(x)
+    # TODO:
+    # Use some more fine criterion, once JAX issue #9567 is fixed.
+    #
+    # if isinstance(x, jax.ad.JVPTracer):
+    #     return True
+    # elif isinstance(x, jax.core.Tracer):
+    #     return any(is_perturbed(attr) for name, attr in x._contents())
+    # else:
+    #     return False
 
 
 def nondifferentiable_input(x: PyTree, name: str) -> None:

--- a/diffrax/misc/errors.py
+++ b/diffrax/misc/errors.py
@@ -24,7 +24,7 @@ def error_if(
         @jax.jit
         def f(x):
             error_if(x < 0, "x must be >= 0")
-            
+
         f(jax.numpy.array(-1))
     """
     branched_error_if(pred, 0, [msg], error_cls)

--- a/diffrax/nonlinear_solver/base.py
+++ b/diffrax/nonlinear_solver/base.py
@@ -8,7 +8,7 @@ import jax.numpy as jnp
 import jax.scipy as jsp
 
 from ..custom_types import Int, PyTree, Scalar
-from ..misc import fixed_custom_jvp, is_perturbed
+from ..misc import fixed_custom_jvp
 from ..solution import RESULTS
 
 
@@ -80,7 +80,9 @@ class AbstractNonlinearSolver(eqx.Module):
         whether the solver managed to converge or not.
         """
 
-        diff_args, nondiff_args = eqx.partition(args, is_perturbed)
+        # TODO: switch from is_inexact_array to is_perturbed once JAX issue #9567 is
+        # fixed.
+        diff_args, nondiff_args = eqx.partition(args, eqx.is_inexact_array)
         return self._solve(self, fn, x, jac, nondiff_args, diff_args)
 
     @staticmethod

--- a/diffrax/solver/heun.py
+++ b/diffrax/solver/heun.py
@@ -1,7 +1,7 @@
 import numpy as np
 
 from ..local_interpolation import ThirdOrderHermitePolynomialInterpolation
-from .base import AbstractStratonovichSolver
+from .base import AbstractAdaptiveSDESolver, AbstractStratonovichSolver
 from .runge_kutta import AbstractERK, ButcherTableau
 
 
@@ -13,7 +13,7 @@ _heun_tableau = ButcherTableau(
 )
 
 
-class Heun(AbstractERK, AbstractStratonovichSolver):
+class Heun(AbstractERK, AbstractStratonovichSolver, AbstractAdaptiveSDESolver):
     """Heun's method.
 
     2nd order explicit Runge--Kutta method. Has an embedded Euler method for adaptive

--- a/diffrax/solver/implicit_euler.py
+++ b/diffrax/solver/implicit_euler.py
@@ -1,7 +1,6 @@
 from typing import Tuple
 
 import jax
-import jax.flatten_util as fu
 
 from ..custom_types import Bool, DenseInfo, PyTree, Scalar
 from ..local_interpolation import LocalLinearInterpolation
@@ -17,10 +16,7 @@ _SolverState = None
 
 def _implicit_relation(z1, nonlinear_solve_args):
     vf_prod, t1, y0, args, control = nonlinear_solve_args
-    _, unravel = fu.ravel_pytree(y0)
-    z1 = unravel(z1)
     diff = (vf_prod(t1, (y0**ω + z1**ω).ω, args, control) ** ω - z1**ω).ω
-    diff, _ = fu.ravel_pytree(diff)
     return diff
 
 
@@ -49,14 +45,13 @@ class ImplicitEuler(AbstractImplicitSolver):
         del solver_state, made_jump
         control = terms.contr(t0, t1)
         pred = terms.vf_prod(t0, y0, args, control)
-        pred, unravel = fu.ravel_pytree(pred)
         jac = self.nonlinear_solver.jac(
             _implicit_relation, pred, (terms.vf_prod, t1, y0, args, control)
         )
         nonlinear_sol = self.nonlinear_solver(
             _implicit_relation, pred, (terms.vf_prod, t1, y0, args, control), jac
         )
-        z1 = unravel(nonlinear_sol.root)
+        z1 = nonlinear_sol.root
         y1 = (y0**ω + z1**ω).ω
         dense_info = dict(y0=y0, y1=y1)
         return y1, None, dense_info, None, nonlinear_sol.result

--- a/diffrax/solver/midpoint.py
+++ b/diffrax/solver/midpoint.py
@@ -1,7 +1,7 @@
 import numpy as np
 
 from ..local_interpolation import ThirdOrderHermitePolynomialInterpolation
-from .base import AbstractStratonovichSolver
+from .base import AbstractAdaptiveSDESolver, AbstractStratonovichSolver
 from .runge_kutta import AbstractERK, ButcherTableau
 
 
@@ -13,7 +13,7 @@ _midpoint_tableau = ButcherTableau(
 )
 
 
-class Midpoint(AbstractERK, AbstractStratonovichSolver):
+class Midpoint(AbstractERK, AbstractStratonovichSolver, AbstractAdaptiveSDESolver):
     """Midpoint method.
 
     2nd order explicit Runge--Kutta method. Has an embedded Euler method for adaptive

--- a/diffrax/solver/ralston.py
+++ b/diffrax/solver/ralston.py
@@ -1,6 +1,7 @@
 import numpy as np
 
 from ..local_interpolation import ThirdOrderHermitePolynomialInterpolation
+from .base import AbstractAdaptiveSDESolver, AbstractStratonovichSolver
 from .runge_kutta import AbstractERK, ButcherTableau
 
 
@@ -21,7 +22,7 @@ _ralston_tableau = ButcherTableau(
 )
 
 
-class Ralston(AbstractERK):
+class Ralston(AbstractERK, AbstractStratonovichSolver, AbstractAdaptiveSDESolver):
     """Ralston's method.
 
     2nd order explicit Runge--Kutta method. Has an embedded Euler method for adaptive
@@ -33,3 +34,6 @@ class Ralston(AbstractERK):
 
     def order(self, terms):
         return 2
+
+    def strong_order(self, terms):
+        return 0.5

--- a/diffrax/solver/runge_kutta.py
+++ b/diffrax/solver/runge_kutta.py
@@ -3,7 +3,6 @@ from dataclasses import dataclass, field
 from typing import Optional, Tuple
 
 import jax
-import jax.flatten_util as fu
 import jax.lax as lax
 import jax.numpy as jnp
 import numpy as np
@@ -96,13 +95,10 @@ def _implicit_relation(ki, nonlinear_solve_args):
     # https://github.com/SciML/DiffEqDevMaterials/blob/master/newton/output/main.pdf
     # (Bearing in mind that our ki is dt times smaller than theirs.)
     diagonal, vf_prod, ti, yi_partial, args, control = nonlinear_solve_args
-    _, unravel = fu.ravel_pytree(yi_partial)
-    ki = unravel(ki)
     diff = (
         vf_prod(ti, (yi_partial**ω + diagonal * ki**ω).ω, args, control) ** ω
         - ki**ω
     ).ω
-    diff, _ = fu.ravel_pytree(diff)
     return diff
 
 
@@ -239,7 +235,6 @@ class AbstractRungeKutta(AbstractAdaptiveSolver):
                 ki_pred = terms.vf_prod(ti, yi_partial, args, control)
             else:
                 ki_pred = vector_tree_dot(self.tableau.a_predictor[i - 1], ω(k)[:i].ω)
-            ki_pred, unravel = fu.ravel_pytree(ki_pred)
             if self._recompute_jac(i):
                 jac = self.nonlinear_solver.jac(
                     _implicit_relation,
@@ -253,7 +248,7 @@ class AbstractRungeKutta(AbstractAdaptiveSolver):
                 (diagonal, terms.vf_prod, ti, yi_partial, args, control),
                 jac,
             )
-            ki = unravel(nonlinear_sol.root)
+            ki = nonlinear_sol.root
             return ki, jac, nonlinear_sol.result
 
 

--- a/test/test_integrate.py
+++ b/test/test_integrate.py
@@ -8,6 +8,7 @@ import jax.numpy as jnp
 import jax.random as jrandom
 import pytest
 import scipy.stats
+from diffrax.misc import ω
 
 from helpers import all_ode_solvers, random_pytree, shaped_allclose, treedefs
 
@@ -507,16 +508,26 @@ def test_compile_time_steps():
     assert shaped_allclose(sol.stats["compiled_num_steps"], jnp.array([20, 20]))
 
 
-def test_grad_implicit_solve():
+@pytest.mark.parametrize("solver", [diffrax.ImplicitEuler(), diffrax.Kvaerno5()])
+def test_grad_implicit_solve(solver):
     # Check that we work around JAX issue #9374
+    # Whilst we're at -- for efficiency -- check the use of PyTree-valued state with
+    # implicit solves.
 
-    term = diffrax.ODETerm(lambda t, y, args: args * y)
-    solve = diffrax.Kvaerno5()
+    term = diffrax.ODETerm(lambda t, y, args: (-args * y**ω).ω)
 
+    @jax.jit
     def f(args):
-        return jnp.sum(
-            diffrax.diffeqsolve(term, solve, t0=0, t1=1, dt0=0.1, y0=1.0, args=args).ys
-        )
+        y0 = (1.0, {"a": 2.0})
+        ys = diffrax.diffeqsolve(term, solver, t0=0, t1=1, dt0=0.1, y0=y0, args=args).ys
+        return jnp.sum(ys[0] + ys[1]["a"])
 
-    grads = jax.grad(f)(1.0)
+    grads = jax.jit(jax.grad(f))(1.0)
     assert jnp.isfinite(grads)
+
+    # Test numerical gradients: Diffrax issue #64
+    eps = 1e-6
+    val = f(1.0)
+    val_eps = f(1.0 + eps)
+    numerical_grads = (val_eps - val) / eps
+    assert shaped_allclose(grads, numerical_grads)

